tensorboard/plugins/graph/graphs_demo.py

```diff
@@ -0,0 +1,131 @@
+# limitations under the License.
+# ==============================================================================
+"""Sample data for the graph plugin.
+
+Most demos emit basic run graph data, but the graph plugin also shows
+more specialized data types. See function docstrings for details about
+what runs have what data.
+"""
+
+import os
+
+import tensorflow as tf
+import numpy as np
+
+LOGDIR = "/tmp/graphs_demo"
+
+
+def main():
+    tagged()
+    profile()
+    keras()
+
+
+def tagged():
+    """Create run graph data with `TaggedRunMetadata`.
+
+    The `tagged` run has a top-level run graph as well as steps
+    `step0000` through `step0002`, each with profile data.
+    """
+    logdir = os.path.join(LOGDIR, "tagged")
+    with tf.compat.v1.Graph().as_default():
+        with tf.compat.v1.Session() as sess:
+            step_tensor = tf.compat.v1.placeholder(shape=(), dtype=tf.int32)
+            output = step_tensor * 2
+
+            writer = tf.compat.v1.summary.FileWriter(logdir)
+            with writer:
+                writer.add_graph(sess.graph)
+                for step in range(3):
+                    feed_dict = {step_tensor: step}
+                    run_options = tf.compat.v1.RunOptions()
+                    run_options.trace_level = tf.compat.v1.RunOptions.FULL_TRACE
+                    run_metadata = tf.compat.v1.RunMetadata()
+                    s = sess.run(
+                        output,
+                        feed_dict=feed_dict,
+                        options=run_options,
+                        run_metadata=run_metadata,
+                    )
+                    writer.add_run_metadata(run_metadata, "step_%04d" % step)
+
+
+def keras():
+    """Create a Keras conceptual graph and op graphs.
+
+    The `keras/train` run has a run-level graph, a `batch_2` tag with op
+    graph only (`graph_run_metadata_graph` plugin), and a `keras` tag
+    with a Keras conceptual graph only (`graph_keras_model` plugin).
+    """
+    logdir = os.path.join(LOGDIR, "keras")
+
+    data_size = 1000
+    train_fac = 0.8
+    train_size = int(data_size * train_fac)
+    x = np.linspace(-1, 1, data_size)
+    np.random.shuffle(x)
+    y = 0.5 * x + 2 + np.random.normal(0, 0.05, (data_size,))
+    (x_train, y_train) = x[:train_size], y[:train_size]
+    (x_test, y_test) = x[train_size:], y[train_size:]
+
+    layers = [
+        tf.keras.layers.Dense(16, input_dim=1),
+        tf.keras.layers.Dense(1),
+    ]
+    model = tf.keras.models.Sequential(layers)
+    model.compile(
+        loss=tf.keras.losses.mean_squared_error,
+        optimizer=tf.keras.optimizers.SGD(lr=0.2),
+    )
+    model.fit(
+        x_train,
+        y_train,
+        batch_size=train_size,
+        verbose=0,
+        epochs=100,
+        validation_data=(x_test, y_test),
+        callbacks=[tf.keras.callbacks.TensorBoard(logdir)],
+    )
+
+
+def profile():
+    """Create data with op graphs and profile data.
```

Comment:
I find it tiny bit odd that the graph_demo is creating profile which is a separate data structure used by a different plugin. If you expect to see profile information in RunMetadata with `profile=True`, I believe the new profiler is separate from RunMetadata and will be useless (it creates profile trace files but do not populate the RunMetadata like L54). Am I missing something?

Reply:
This is the only way that I could find to generate data that lives under
the `graph_run_metadata` plugin. It’s written in an unexported function,
`summary_ops_v2.run_metadata`, which [looks to only be called in this
code path][1].

Is there a different way that I can test [this functionality][2]?

[1]: https://cs.opensource.google/search?q=%5Cbrun_metadata%5C(%20f:%5Etensorflow%2F%20lang:py%20-f:_test%5C.py$
[2]: https://github.com/tensorflow/tensorboard/blob/945f39065ff25827b7d69924444e81d51f8e059a/tensorboard/plugins/graph/graphs_plugin.py#L156-L158

Reply:
I see. For the `trace_on(graph=True, profiler=False)`, you are relying on the Keras callback? Since we do not closely control the TB Keras callback, I think it makes more sense to explicitly exercise https://github.com/tensorflow/tensorboard/blob/945f39065ff25827b7d69924444e81d51f8e059a/tensorboard/plugins/graph/graphs_plugin.py#L47 flow, too :)

Reply:
Yeah, that’s right. Sure, I suppose there’s no harm in doing that.